### PR TITLE
Mediactive login

### PR DESCRIPTION
**Pull request by maparent**
New social-auth option to work with mediactive. Tests TBD.
Many changes to social login pipeline, because we cannot assume the user came to the social login through assembl; so discussion or user may no be set, or may not be correct.

### CODE REVIEW

File: assembl/auth/encryption.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+class AESDecryptor(Decryptor):
+    """A decryptor that uses AES symmetric keys"""
+    def __init__(self, password, backend=None):
+        backend = backend or default_backend()
+        self.backend = backend
+        self.password = password
+
+    def decrypt(self, message):
+        iv = base64.b64decode(message[:24])
```
Comment (ayyazdaniaryan) [inline assembl/auth/encryption.py]: `24` should be a constant with a comment for maintainability.

File: assembl/auth/encryption.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        cipher = Cipher(
+            algorithms.AES(self.password),
+            modes.CBC(iv), backend=self.backend)
+        message = base64.b64decode(message[24:])
+        decryptor = cipher.decryptor()
+        padded = decryptor.update(message)
+        padded += decryptor.finalize()
+        unpadder = padding.PKCS7(128).unpadder()
```
Comment (ayyazdaniaryan) [inline assembl/auth/encryption.py]: So should the padder int.

File: assembl/auth/encryption.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+class AESEncryptor(Encryptor):
+    """An encryptor that uses AES symmetric keys"""
+    def __init__(self, password, backend=None):
+        backend = backend or default_backend()
+        self.backend = backend
+        self.password = password
+
+    def encrypt(self, message):
+        iv = urandom(16)
```
Comment (ayyazdaniaryan) [inline assembl/auth/encryption.py]: The bit counts should also be const. Likely as a member variable on the `Encryptor` root class.

File: assembl/auth/mediactive.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    name = 'mediactive'
+    USERNAME_KEY = 'username'
+    html_parser = HTMLParser()
+
+    def auth_url(self):
+        """Must return redirect URL to auth provider"""
+        next_url = self.data.get('next', None)
+        base = self.setting('LOGIN_URL')
+        if next_url:
+            base += '?' + urlencode({'next': next_url})
```
Comment (ayyazdaniaryan) [inline assembl/auth/mediactive.py]: Use `urljoin` and `urlencode` to ensure flexible usage of the config file under the `LOGIN_URL` key.

Comment (maparent) [reply assembl/auth/mediactive.py]: urlencode I'm using... not sure what urljoin is adding anything to this. 
Unless you mean urlparse/urldefrag?

Comment (maparent) [reply assembl/auth/mediactive.py]: I wrote it for the heck of it, but I think it's overkill. Tell me if you want it.
```python
        if next_url:
            pr = urlparse(base)
            qs = parse_qs(pr.query)
            qs['next'] = [next_url]
            base = urlunparse(pr._replace(query=urlencode(qs, True)))
```

**Pull request merged**